File: 1044-find-common-characters/find-common-characters.java
class Solution {
    public List<String> commonChars(String[] words) {
        HashMap <Character, Integer> map = new HashMap <> ();
        for(char ch : words[0].toCharArray()){
            map.put(ch, map.getOrDefault(ch , 0) + 1);
        }

        for(int i = 1 ; i < words.length; i++){
            HashMap <Character, Integer> freq = new HashMap <> ();
            for(char ch: words[i].toCharArray()){
                freq.put(ch, freq.getOrDefault(ch , 0) + 1);
            }

            for(char ch : map.keySet()){
                if(freq.containsKey(ch)){
                    map.put(ch, Math.min(map.get(ch), freq.get(ch)));
                }
                else map.put(ch, 0);
            }
        }

        List <String> ans = new ArrayList <>();
        for(char ch : map.keySet()){
            int count = map.get(ch);
            while(count > 0){
                ans.add(String.valueOf(ch));
                count--;
            }
        }
        return ans;   
    }
}

File: 1044-find-common-characters/find-common-characters.cpp
class Solution {
public:
    vector<string> commonChars(vector<string>& words) {
        vector <int> map(26);
        for(char ch : words[0]){
            map[ch - 'a']++;
        }

        for(int i = 1 ; i < words.size(); i++){
            vector <int> freq (26);
            for(char ch: words[i]){
                freq[ch - 'a']++;
            }

            for(int j = 0 ; j < 26; j++){
                map[j] = min(map[j], freq[j]);
            }
        }

        vector<string> ans;
        for(int i = 0 ; i < 26; i++){
            int count = map[i];
            while(count > 0){
                ans.push_back(string(1, (char)(i + 'a')));
                count--;
            }
        }
        return ans;
    }
};